src/eh1/delay.rs

```diff
@@ -1,23 +1,403 @@
+/// Nanoseconds per millisecond
+const NANOS_PER_MILLI: u32 = 1_000_000;
+/// Microseconds per millisecond
+const MICROS_PER_MILLI: u32 = 1_000;
+
+impl Transaction {
+    /// Create a new delay transaction
+    pub fn new(kind: TransactionKind) -> Transaction {
+        Transaction {
+            kind,
+            real_delay: false,
+        }
+    }
+
+    /// Create a new delay_ns transaction
+    pub fn delay_ns(ns: u32) -> Transaction {
+        Transaction::new(TransactionKind::DelayNs(ns))
+    }
+
+    /// Crete a new delay_us transaction
+    pub fn delay_us(us: u32) -> Transaction {
+        Transaction::new(TransactionKind::DelayUs(us))
+    }
+
+    /// Create new delay_ms transaction
+    pub fn delay_ms(ms: u32) -> Transaction {
+        Transaction::new(TransactionKind::DelayMs(ms))
+    }
+
+    /// Crete a new async delay_ns transaction
+    #[cfg(feature = "embedded-hal-async")]
+    pub fn async_delay_ns(ns: u32) -> Transaction {
+        Transaction::new(TransactionKind::AsyncDelayNs(ns))
+    }
+
+    /// Crete a new async delay_us transaction
+    #[cfg(feature = "embedded-hal-async")]
+    pub fn async_delay_us(us: u32) -> Transaction {
+        Transaction::new(TransactionKind::AsyncDelayNs(us * 1_000))
+    }
+
+    /// Crete a new async delay_ms transaction
+    #[cfg(feature = "embedded-hal-async")]
+    pub fn async_delay_ms(ms: u32) -> Transaction {
+        Transaction::new(TransactionKind::AsyncDelayNs(ms * 1_000_000))
+    }
+
+    /// Crete a new async delay_ns transaction
+    #[cfg(feature = "embedded-hal-async")]
+    pub fn any_delay_ns(ns: u32) -> Transaction {
+        Transaction::new(TransactionKind::AnyDelayNs(ns))
+    }
+
+    /// Crete a new async delay_us transaction
+    #[cfg(feature = "embedded-hal-async")]
+    pub fn any_delay_us(us: u32) -> Transaction {
+        Transaction::new(TransactionKind::AnyDelayNs(us * 1_000))
+    }
+
+    /// Crete a new async delay_ms transaction
+    #[cfg(feature = "embedded-hal-async")]
+    pub fn any_delay_ms(ms: u32) -> Transaction {
+        Transaction::new(TransactionKind::AnyDelayNs(ms * 1_000_000))
+    }
+
+    /// Perform an actual delay for this transaction
+    pub fn wait(mut self) -> Transaction {
+        self.real_delay = true;
+        self
+    }
+}
+
+/// MockDelay transaction kind.
+#[derive(PartialEq, Eq, Clone, Debug)]
+pub enum TransactionKind {
+    /// Delay in nanoseconds
+    DelayNs(u32),
+    /// Asynchronous delay in nanoseconds
+    AsyncDelayNs(u32),
+    /// Any delay in nanoseconds, blocking or async
+    AnyDelayNs(u32),
+    /// Delay in microseconds
+    DelayUs(u32),
+    /// Asynchronous delay in microseconds
+    AsyncDelayUs(u32),
+    /// Any delay in microseconds, blocking or async
+    AnyDelayUs(u32),
+    /// Delay in milliseconds
+    DelayMs(u32),
+    /// Asynchronous delay in milliseconds
+    AsyncDelayMs(u32),
+    /// Any delay in milliseconds, blocking or async
+    AnyDelayMs(u32),
+}
```

Comment:
Why do we have Delay vs AsyncDelay vs AnyDelay?

If I understand it correctly, the Delay* variants can only deal with sync calls, the AsyncDelay* variants can only deal with async calls, and AnyDelay* forwards calls to both traits.

However, in which situation would one want a mock object that only supports one or the other variant? Couldn't we just implement both interfaces (as long as the necessary Cargo features are enabled) and call it a day?

Reply:
Interface implementation is something else, as this is done for a CheckedDelay struct and not a transaction.

The idea behind differentiation between blocking and async behavior is to allow tests to validate that async functions properly use delays.
E.g.
In a big async function, there should be 100ms delay. If the intention of the author is to use non-blocking delay, they can specify so in their tests. Now, if they include some external dependency or call some functions via polymorphic trait, tests would fail for a 100ms blocking delay. This allows for more specific tests, where some calls are supposed to be short and blocking and some should be long and async.

I wouldn't remove this functionality, but if this is an uncommon scenario and user-friendliness is of a big concern, then maybe these transactions can be renamed from:
-  Delay*
-  AsyncDelay*
-  AnyDelay*

to:
-  Delay* (blocking or async, don't care. sane default for 99% cases)
-  AsyncDelay* (must be async)
-  BlockingDelay* (must be blocking)

Reply:
I see! Yeah, I think this is quite a specialized use case. However, with your rename suggestion, I think that would already be a big improvement.

What's your opinion, @newAM?

Reply:
I think the rename is a good idea, with maybe a note in the docstrings of each that `Delay` should be used in almost all cases.

Reply:
@dbrgn @newAM I've updated docstrings to state, that Delay is a preferred default for most use cases.
Also, the examples shown on top of the module use Delay.